labgrid/util/ssh.py

```diff
@@ -0,0 +1,324 @@
+
+    def get(self, host: str):
+        """Retrieve or create a new connection to a given host
+
+        Arguments:
+            host (str): host to retrieve the connection for
+
+        Returns:
+            :obj:`SSHConnection`: the SSHConnection for the host"""
+        host = socket.getfqdn(host)
+        instance = self._connections.get(host)
+        if instance is None:
+            # pylint: disable=unsupported-assignment-operation
+            self.logger.debug("Trying to start new control socket")
+            # instance = self._start_control_socket(host)
+            instance = SSHConnection(host)
+            instance.connect()
+            self._connections[host] = instance
+        return instance
+
+    def add_connection(self, connection):
+        """Add an existing SSHConnection to the manager.
+        This is useful for manually created connections which should also be
+        available via the manager.
+
+        Arguments:
+            connection (:obj:`SSHConnection`): SSHconnection to add to the manager
+        """
+        # pylint: disable=unsupported-assignment-operation
+        assert isinstance(connection, SSHConnection)
+        if connection.host not in self._connections:
+            self._connections[connection.host] = connection
+
+    def remove_connection(self, connection):
+        # pylint: disable=unsupported-assignment-operation
+        assert isinstance(connection, SSHConnection)
+        if connection.isconnected():
+            raise ExecutionError("Can't remove connected connection")
+        del self._connections[connection.host]
+
+    def remove_by_name(self, name):
+        # pylint: disable=unsupported-assignment-operation
+        del self._connections[name]
+
+    def open(self, host):
+        return self.get(host)
+
+    def close(self, host):
+        con = self.get(host)
+        con.disconnect()
+        self.remove_connection(con)
+
+    def request_forward(self, host, dest, port):
+        con = self.get(host)
+        return con.add_port_forward(dest, port)
+
+    def remove_forward(self, host, dest, port):
+        con = self.get(host)
+        con.remove_port_forward(dest, port)
+
+    def put_file(self, host, local_file, remote_file):
+        con = self.get(host)
+        con.put_file(local_file, remote_file)
+
+    def get_file(self, host, remote_file, local_file):
+        con = self.get(host)
+        con.get_file(remote_file, local_file)
+
+    def close_all(self):
+        """Close all open connections and remove them from the manager """
+        for name, connection in self._connections.items():
+            if connection.isconnected():
+                connection.disconnect()
+        names = self._connections.copy().keys()
+        for name in names:
+            self.remove_by_name(name)
+
+
+def check_connected(func):
+    """Check if an SSHConnection is connected as a decorator"""
+
+    @wraps(func)
+    def wrapper(cls, *_args, **_kwargs):
```

Comment:
Why do you use cls instead of self? The value should be a object, not the class.

Reply:
I'll rename `cls` to `obj`

Reply:
See the BindingMixin.check_active decorator. You can do the sme here.